### ISSUE TITLE
Is this SDK affected by the CallStranger vulnerability?

### ISSUE THREAD

**Issue by StudioMaX** [pupnp/pupnp#180]
I am not a user of this library, but it is used in software for which a vulnerability test showed a positive result.
Is this library affected in any way by the [CallStranger](https://callstranger.com/) vulnerability (CVE-2020-12695)?

**Comment by ffontaine** [comment]
Yes, it is vulnerable as `DeliveryURLs` are not checked in `gena_process_subscription_request`. This was not requested by the UPnP UDA until https://openconnectivity.org/upnp-specs/UPnP-arch-DeviceArchitecture-v2.0-20200417.pdf. Here is an extract of the new requirements:

> The subscription request containing a delivery URL not on the same network segment as the fully qualified event subscription URL shall not  be accepted. For private networks this means that the delivery URL provided will adhere to the following IP ranges:
> - 10.0.0.0 - 10.255.255.255 (10/8 prefix)
> - 172.16.0.0 - 172.31.255.255 (172.16/12 prefix)
> - 192.168.0.0 - 192.168.255.255 (192.168/16 prefix)

So, pupnp will have to be updated to be compliant with this new requirement. I can try to cook something but it will have to be reviewed/updated by more people. I think that the best course of action would be to retrieve the netmask and not just the IP address in UPnPInit2. UPnPInit should be removed as it is deprecated for a long time or updated so the user can also pass the netmask. The netmask should be retrieved for IPv4 but also IPv6, Windows and Linux.

**Comment by ffontaine** [comment]
@mrjimenez, you'll find a PR for this issue here: https://github.com/pupnp/pupnp/pull/181. There is still some work to do as explained in the commit message but could you review this first iteration?

**Comment by mrjimenez** [comment]
Looks ok to me.

If someone wants to test, there is this tester script on github: https://github.com/yunuscadirci/CallStranger.

Regards,
Marcelo.

**Comment by ffontaine** [comment]
Yes, this script is very useful and works very well. I'll try to work on the IPv6 part but I have some difficulties to make it work again and the CallStranger script does not manage IPv6. We'll also need some support from Windows users to retrieve the netmask on this platform and to be sure that nothing was broken with my first PR.

**Comment by ffontaine** [comment]
I think I fixed the issue on IPv4 and IPv6 for "standard" linux as well as BSD-flavors. I wasn't able to fix it for Windows as I'm not a Windows developer.

**Comment by mrjimenez** [comment]
I don't know if it is currently possible to easily build libupnp under windows.

**Comment by mrjimenez** [comment]
Now that we can build on Windows, is there anyone who could test agains CallStranger so that we can close this issue?
